control-plane/connect-inject/controllers/endpointsv2/write_cache_test.go

```diff
@@ -0,0 +1,237 @@
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, randomString(), args.k8sUid)
+			},
+			want: false,
+		},
+		{
+			name: "Non-matching k8sUid returns false",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), randomString())
+			},
+			want: false,
+		},
+		{
+			name: "Matching data returns true",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+			},
+			want: true,
+		},
+		{
+			name: "Removed data returns false",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+				cache.update("another-key", randomBytes(), randomString(), randomString())
+				cache.remove(args.key)
+			},
+			want: false,
+		},
+		{
+			name: "Replaced data returns false",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+				cache.update(args.key, randomBytes(), args.generationFetchFn(), args.k8sUid)
+			},
+			want: false,
+		},
+		{
+			name: "Invalid hash does not update cache",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+				cache.update(args.key, []byte{}, args.generationFetchFn(), args.k8sUid)
+			},
+			want: true,
+		},
+		{
+			name: "Invalid generation does not update cache",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+				cache.update(args.key, args.hash, "", args.k8sUid)
+			},
+			want: true,
+		},
+		{
+			name: "Invalid k8sUid does not update cache",
+			args: args{
+				"foo",
+				testHash,
+				func() string {
+					return testGeneration
+				},
+				testK8sUid,
+			},
+			setupFn: func(args args, cache WriteCache) {
+				cache.update(args.key, args.hash, args.generationFetchFn(), args.k8sUid)
+				cache.update(args.key, args.hash, args.generationFetchFn(), "")
+			},
+			want: true,
+		},
```

Comment:
Not sure what I would expect from the behavior here if someone wants to update the cache with junk values. In that case, I think we'd want to fail on the side of always wanting to invalidate it. No strong opinion here.

Reply:
That's a fair callout. I think here it will be 6+half-dozen due to the fact that either way, the next write is guaranteed (we didn't match and wrote to Consul, hence the cache write - so skipping the cache update should be equivalent to invalidating). I expect the only solution in either case will be to notice the error logs or tons of writes from the controller and investigate.